### PR TITLE
Fixes #283

### LINKED ISSUE #283: [BUG] Auto slicing is not supported for DeleteByQuery, Reindex, and UpdateByQuery
### What is the bug?
The API does not allow defining auto-slicing for `DeleteByQueryRequest`, `ReindexRequest`, and `UpdateByQueryRequest`.
JavaDoc suggests using 'auto' but only `Long` value is accepted.

```
		/**
		 * The number of slices this task should be divided into. Defaults to 1, meaning
		 * the task isn't sliced into subtasks. Can be set to <code>auto</code>.
		 * <p>
		 * API name: {@code slices}
		 */
		public final Builder slices(@Nullable Long value) {
			this.slices = value;
			return this;
		}

```
### How can one reproduce the bug?
n/a

### What is the expected behavior?
Add support for configuring automatic slicing.

### What is your host/environment?
Linux/Windows, Java 17
OpenSearch server 2.4.0
OpenSearch client 2.1.0

### Do you have any screenshots?
n/a

### PR DESCRIPTION
**Pull request by urinud**
### Description
Adds the option to set slices=auto for UpdateByQueryRequest, DeleteByQueryRequest and ReindexRequest.
Using value 0 (zero) to express auto slicing.
This maintains consistency with https://github.com/elastic/elasticsearch/pull/53068

### Issues Resolved
Fixes #283 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
Thanks for the pull request @urinud , could you please add a few test cases to verify the change under `AbstractRequestIT` (or other appropriate test suite)? thank you.

**Comment by dblock** [comment]
Also this is a breaking change to existing users, isn't it?

**Comment by urinud** [comment]
> Also this is a breaking change to existing users, isn't it?

I don’t think is a breaking change as zero is not a valid value for the slices. Also this change maintains consistency with the equivalent feature in Elastic.

**Review by VachaShah — COMMENTED** [review]
@urinud Can you add a changelog entry in [CHANGELOG](https://github.com/opensearch-project/opensearch-java/blob/main/CHANGELOG.md) for this PR?

File: java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
```
Comment (dblock) [inline java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java:9]: For new files you don't need to include it. 

Please also confirm that you did not copy this from non-APLv2 code.

File: java-client/src/test/java/org/opensearch/client/opensearch/core/ReindexRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
```
Comment (dblock) [inline java-client/src/test/java/org/opensearch/client/opensearch/core/ReindexRequestTest.java:9]: Same as above.

**Review by dblock — CHANGES_REQUESTED** [review]
Looks good. 

Please remove ES license from new files, and confirm that you're not copying any code from non-open-source (APLv2-compatible) codebases.

For DCO I think you need to amend your commits. LMK if you need help with that?

File: java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java (hunk 1 above)
Comment (urinud) [reply java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java]: Removed it. Thanks

File: java-client/src/test/java/org/opensearch/client/opensearch/core/ReindexRequestTest.java (hunk 2 above)
Comment (urinud) [reply java-client/src/test/java/org/opensearch/client/opensearch/core/ReindexRequestTest.java]: Removed it. Thanks.

**Review by dblock — APPROVED** [review]

File: java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
```
Comment (dblock) [inline java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java:4]: Sorry if I wasn't clear. The correct header is as follows:

```
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 */
```

Just copy it from other files.

File: CHANGELOG.md [hunk 4]
```diff
@@ -32,6 +32,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ### Dependencies
 
 ### Changed
+- Adds the option to set slices=auto for UpdateByQueryRequest, DeleteByQueryRequest and ReindexRequest ([#538](https://github.com/opensearch-project/opensearch-java/pull/538))
```
Comment (dblock) [inline CHANGELOG.md:35]: I think this is a new feature, let's move it to "Added"?

**Review by dblock — CHANGES_REQUESTED** [review]

File: java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java (hunk 3 above)
Comment (urinud) [reply java-client/src/test/java/org/opensearch/client/opensearch/core/DeleteByQueryRequestTest.java:7]: Done.

File: CHANGELOG.md (hunk 4 above)
Comment (urinud) [reply CHANGELOG.md]: Done

**Review by dblock — APPROVED** [review]

**Comment by urinud** [comment]
Hello @reta and @VachaShah can you please help reviewing the PR? Is there anything else I should have added?
Thank you

**Review by reta — APPROVED** [review]

**Pull request merged**